util/feed.go

```diff
@@ -0,0 +1,396 @@
+//
+// The Feed will continue running until its associated Stopper is closed.
+//
+//		func FeedExample() {
+//			stopper := NewStopper()
+//			feed := StartFeed(stopper)
+//			for i := 0; i < 5; i++ {
+//				go StartSubscriber(i, feed)
+//			}
+//			feed.Publish("Event 1"
+//			feed.Publish("Event 2")
+//			stopper.Stop()
+//		}
+//
+//		func StartSubscriber(i int, f *Feed) {
+//			sub := f.Subscribe()
+//			for event := range sub.Events {
+//				fmt.Println("Subscriber %d received event %s", i,
+//					event.(string))
+//			}
+//		}
+//
+// The Feed's non-blocking properties are achieved with a linked channel instead
+// of buffered channels, and thus there is no limit (other than available
+// memory) to how many unread events the Feed can maintain. This is desirable
+// for Feeds which may publish data in quick bursts.
+//
+// The Feed does not keep historical events; individual Subscribers will only
+// receive events published after they subscribe.
+type Feed struct {
+	events  chan interface{}
+	stopper *Stopper
+
+	// subscriberCount counts the total number of subscribers that have
+	// ever been started. activeCount counts the number of subscribers still
+	// active (i.e. that have not been unsubscribed).
+	subscriberCount int
+	activeCount     int
+
+	// these channels are used for synchronizing different subscription
+	// lifecycle tasks.
+	subscribe   chan *publisher
+	unsubscribe chan *publisher
+	closer      chan struct{}
+}
+
+// StartFeed instatiates and starts a new Feed. The Feed can immediately accept
+// Subscribers and publish events.
+func StartFeed(stopper *Stopper) *Feed {
+	f := &Feed{
+		events:      make(chan interface{}),
+		stopper:     stopper,
+		subscribe:   make(chan *publisher),
+		unsubscribe: make(chan *publisher),
+		closer:      make(chan struct{}),
+	}
+	go func() {
+		var (
+			// head is the most recent entry added to the feed. This is also the
+			// current head of the linked channel of entries. The first entry
+			// will never be consumed by any subscribers, but initial
+			// subscribers may listen on its "next" channel.
+			head = feedEntryPool.Get().(*feedEntry)
+			// addEntry is used to correctly add a new entry to the linked channel.
+			addEntry = func(et entryType, value interface{}) {
+				entry := feedEntryPool.Get().(*feedEntry)
+				entry.entryType = et
+				entry.consumerCount = f.activeCount
+				switch et {
+				case valueEntry:
+					entry.value = value
+				case unsubEntry:
+					entry.unsubscribedID = value.(int)
+				}
+				head.next <- entry
+				head = entry
+			}
+		)
+		for {
+			select {
+			case pub := <-f.subscribe:
+				// A new Subscription is requested, and we have received the
+				// publisher associated with it. Give the publisher the channel
+				// for the next entry and start it's process() goroutine.
+				f.subscriberCount++
+				f.activeCount++
+				pub.id = f.subscriberCount
+				pub.current = head
+				go pub.process()
```

Comment:
I think this should also be using RunWorker, these are long-running but also should cease when the stopper is closed.